### ISSUE TITLE
Deadlock in thread_queue.h

### ISSUE THREAD

**Issue by majh** [eclipse/paho.mqtt.cpp#227]
I think that there is a deadlock in thread_queue.h due to the queue locking and signalling.
For example, if queue size is 1, and
*  get() is called with empty queue, n == 0, and waits on notEmptyCond
* put runs, adds valuie, singnals notEmptyCond notify_once 
* put runs, wait's on notFullCond
* get wakes, gets value, but doesn't notify notFullCond as n remains 0

Please see commits below for a proposed fix
https://github.com/majh/paho.mqtt.cpp/commit/e24f43fba8959c03084209909b7b594bf5edbe92
https://github.com/majh/paho.mqtt.cpp/commit/6ba94c21d0a1425612758d69d8179c28ebfaf491

**Comment by fpagliughi** [comment]
Hey @majh. You're totally right!!!

At first I thought this was a weird corner case when _capacity_ is 1. Normally, if I need that, I use an optimized structure that just uses a single variable instead of a queue and call the whole thing a _mail_slot_, or something like that. But that's another story.

The problem exists with any capacity, though less likely to be observed the bigger capacity gets. You're right, the problem is relying on a cached value of size as `n` from before the wait. Any number of readers or writers could get in there while a thread is waiting, and upon return, `n` could be anything.

So it needs to be re-checked after returning from a wait.

But I do want to keep the optimization that the condition variables are signalled _only_ when their state changes (from empty to not empty, etc). I don't want to signal every time an item is added or removed, as that could be an expensive system call on some platforms.

So we could get rid of `n` and explicitly check every time:
```
value_type get() {
    unique_guard g(lock_);
    if (queue_.empty())
        notEmptyCond_.wait(g, [=]{return !que_.empty();});
    value_type val = std::move(que_.front());
    que_.pop();
    if (que_.size() == cap_-1) {
        g.unlock();
        notFullCond_.notify_one();
    }
    return val;
}
``` 

Or, we could just update `n` upon return from the wait:
```
value_type get() {
    unique_guard g(lock_);
    auto n = que_.size();
    if (n == 0) {
        notEmptyCond_.wait(g, [=]{return !que_.empty();});
        n = que_.size();
    }
    value_type val = std::move(que_.front());
    que_.pop();
    if (n == cap_) {
        g.unlock();
        notFullCond_.notify_one();
    }
    return val;
}
```
I think either way would work.

**Comment by majh** [comment]
Hi @fpagliughi,
Possibly the  the 2nd option is optimal as it would only involve a single queue size check (size/empty) most of the time as the n is cached ? ( note in second option should be (n == cap - 1)
Then 
```
void put(value_type val) {
        unique_guard g(lock_);
        auto n = que_.size();
        if (que_.size() >= cap_)
        {
            notFullCond_.wait(g, [&]{return que_.size() < cap_;});
            n = que_.size();
        }
        que_.emplace(std::move(val));
        if (n == 1)
        {
            g.unlock();
            notEmptyCond_.notify_one();
        }
    }
```

Also a couple of other queries
1. Would it be an optimisation to pass the lambda capture by reference [&] so that the queue didn't need to be copied, ie by value ?

2. In  pub_speed_test.cpp,  if the thread_queue size is not smaller than the  'maxBufferedMessages'  (100 default) of pahoo.mqtt.c, this can lead to some type of deadlock due to a async_client::publish throwing an exception due to maxBufferedMessages being exceeded, (MQTTAsync_send/ rc = MQTTASYNC_MAX_BUFFERED_MESSAGES), which then suspends the main publishing thread, and then the async waiter remains blocked waiting on the queue. Not sure how this type of behaviour should be documented/avoided, as there is a dependency on the message buffer sizes in the underlying library ?

**Comment by fpagliughi** [comment]
Normally I would say, sure, send a PR over. But I have variations of this file in a bunch of different projects. So I might as well spend a day and check/fix them all. But actually I think/hope this "optimization" of the cached "n" was something I just did here.

**Comment by fpagliughi** [comment]
Fixed. Thanks

**Comment by AlexisTM** [comment]
Shouldn't there be a new release to reflect this critical addition?

**Comment by fpagliughi** [comment]
There is a new release being prepared with a number of bug fixes, missing MQTT v5 features, and some new features.

**Comment by AlexisTM** [comment]
Thanks! That's great news :)
